src/main/java/net/io/kino/service/impl/AuthenticationManagerImpl.java

```diff
@@ -0,0 +1,68 @@
+package net.io.kino.service.impl;
+
+import net.io.kino.repository.ManagersRepository;
+import net.io.kino.model.Manager;
+import net.io.kino.service.AuthenticationManager;
+import org.apache.tomcat.util.codec.binary.Base64;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import javax.crypto.SecretKey;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.spec.InvalidKeySpecException;
+
+public class AuthenticationManagerImpl implements AuthenticationManager {
+
+    private static final int iterations = 20*1000;
+    private static final int saltLen = 32;
+    private static final int desiredKeyLen = 256;
+
+    @Autowired
+    private ManagersRepository managers;
+
+    @Override
+    public boolean login(String username, String password) throws InvalidKeySpecException, NoSuchAlgorithmException {
+        return authenticate(username, password);
+    }
+
+    @Override
+    public boolean logout(Manager user) {
+        return false;
+    }
+
+    // check if hash of a given password corresponds with a password in database
+    @Override
+    public boolean authenticate(String username, String password) throws InvalidKeySpecException, NoSuchAlgorithmException {
+        Manager found = managers.getOne(username);
+        String stored = found.getPassword();
+
+        String[] saltAndHash = stored.split("\\$");
+        if (saltAndHash.length != 2) {
+            throw new IllegalStateException(
+                    "Stored password doesn't have a form 'salt$hash'");
+        }
+        String hashOfInput = hash(password, Base64.decodeBase64(saltAndHash[0]));
+        return hashOfInput.equals(saltAndHash[1]);
+    }
+
+    // compute salted PBKDF2 hash of given password
+    public static String getSaltedHash(String password) throws NoSuchAlgorithmException, InvalidKeySpecException {
+        byte[] salt = SecureRandom.getInstance("SHA1PRNG").generateSeed(saltLen);

```

Comment:
Jeśli chodzi o te wyjątki NoSuchAlgorithmException i InvalidKeySpecException, to chyba bym je tutaj złapał po prostu i rzucił jakiś IllegalState w zamian. 
Używając tych "zarażamy" dalszy kod programu, a i tak nikt z tym nic nie zrobi. Użytkownik tego API nie ma wpływu na to co się dzieje w tej metodzie, więc nie ma temu jak zapobiec.

Reply:
Ok, to tak zrobię